cabby/rvs/observe.py

```diff
@@ -0,0 +1,75 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''Functions to support observation and selection of POIs on paths.'''
+
+from typing import Dict, Sequence, Text
+
+from shapely.geometry.point import Point
+
+from cabby.data.wikidata import item
+from cabby.geo import util
+
+def get_all_distances(
+  focus: Point, entities: Sequence[item.Entity]) -> Dict[Text, float]:
+  """Get the distance from each entity to the provided focus point.
+  
+  Args:
+    focus: The point to compute distances to.
+    entities: The geolocated entities to compute distances from.
+  Returns:
+    A dictionary with the QIDs of the entities as keys and the distances from
+    the corresponding entity to the focus point.
+
+  """
+  distances = {}
+  for entity in entities:
+    distances[entity.qid] = util.get_distance_km(focus, entity.location)
+  return distances
+
+
+def get_pivot_poi(
+  start: Point, goal: Point, entities: Sequence[item.Entity]) -> Text:
+  """Select a POI that can act as a useful reference between start and goal.
+
+  Args:
+    start: The starting point of a route.
+    goal: The goal point of a route.
+    entities: The accessible entities to consider.
+  Returns:
+    The QID of the highest ranked POI linking the start and goal points. 
+    Currently this is determined simply by ranking all entities by whether they
+    are between the start and goal, and preferring entities that are about 
+    three-quarters of the way to the goal from the start. This can be improved
+    considerably in future.
+  """
+  pdist = util.get_distance_km(start, goal)
+  candidate_scores = {}
+  for entity in entities:
+    odist = util.get_distance_km(start, entity.location)
+    ddist = util.get_distance_km(goal, entity.location)
+    if odist < pdist and ddist < pdist and odist+ddist < pdist*1.05:
```

Comment:
same for this. any intuition for the 5% value? would it depend on how far the start and end are? e.g. it should not account just a few steps if they are too close.

Reply:
Agree! Done. (And we might want to make it more like 1.10 -- needs more experimentation. OTOH, this will likely change significantly anyway.